File: 2019/src/day06.dart
import '../../shared/dart/src/utils.dart';

/// Puzzle description: https://adventofcode.com/2019/day/6

const bool DEBUG = false;

//Return number of objects orbited by object
int orbitCount(Map orbits, object) {
  var total = 0;
  var orbitedObject = orbits[object];
  if (orbitedObject != null) total = orbitCount(orbits, orbitedObject) + 1;
  return total;
}

//Return list of objects orbited by object.
List orbitedObjects(Map orbits, object) {
  var total = [];
  var orbitedObject = orbits[object];
  if (orbitedObject != null) {
    total
      ..add(orbitedObject)
      ..addAll(orbitedObjects(orbits, orbitedObject));
  }
  return total;
}

//key = orbiting object, value = orbited object
Map orbitMap(input) => {for (var o in input) o.split(')')[1]: o.split(')')[0]};
//part 1 alternative that uses the method for part 2
int part1Alternative(String header, List input) {
  printHeader(header);
  var orbits = orbitMap(input);
  return orbits.keys
      .fold(0, (sum, o) => sum + orbitedObjects(orbits, o).length);
}

int part1(String header, List input) {
  printHeader(header);
  var orbits = orbitMap(input);
  return orbits.keys.fold(0, (sum, o) => sum + orbitCount(orbits, o));
  // return part1Alternative(header, input);
}

int part2(String header, List input) {
  printHeader(header);
  var orbits = orbitMap(input);
  var youOrbits = orbitedObjects(orbits, 'YOU');
  var sanOrbits = orbitedObjects(orbits, 'SAN');
  var bothOrbit = youOrbits.toSet().intersection(sanOrbits.toSet());
  return (youOrbits.indexOf(bothOrbit.first)) +
      (sanOrbits.indexOf(bothOrbit.first));
}

void main(List<String> arguments) {
  List testInput = FileUtils.asLines('../data/day06-test.txt');
  List testInputb = FileUtils.asLines('../data/day06b-test.txt');
  List mainInput = FileUtils.asLines('../data/day06.txt');

  //Answer:
  assertEqual(part1('06 test part 1', testInput), 42);
  //Answer:
  assertEqual(part2('06 test part 2', testInputb), 4);

  //Answer:171213
  print(part1('06 part 1', mainInput));
  //Answer: 292
  print(part2('06 part 2', mainInput));
}

File: 2023/src/day06.py
import re
import os
import sys

sys.path.append(os.path.relpath("../../shared/python"))

# noqa stops autopep8 from reordering this import
from utils import *  # noqa: E402

# Puzzle description: https://adventofcode.com/2023/day/06


def read_input(input_file):
    input = read_file_str(input_file, True)
    re.findall(r'\d+', input[0])

    times = list(map(int, input[0].split(':')[1].split()))
    distances = list(map(int, input[1].split(':')[1].split()))
    return (times, distances)


#  def ceiling_division(a, b):
    #  return (a + b - 1) // b


def distance_exceeded(v, d, t):
    # from v = d/t
    return (t-v)*v > d


def part1(input):
    times = input[0]
    distances = input[1]
    result = 1

    for i, t in enumerate(times):
        d = distances[i]
        result *= sum(1 for v in range(t) if distance_exceeded(v, d, t))

    return result


def part2(input):
    t = int(''.join(map(str, input[0])))
    d = int(''.join(map(str, input[1])))

    return sum(1 for v in range(t) if distance_exceeded(v, d, t))


def main():
    input = read_input("../data/day06.txt")
    test_input = read_input("../data/day06-test.txt")

    assert (res := part1(test_input)) == 288, f'Actual: {res}'

    print(f'Part 1 {part1(input)}')  # 170000

    assert (res := part2(test_input)) == 71503, f'Actual: {res}'

    print(f'Part 2 {part2(input)}')  # 20537782


if __name__ == '__main__':
    main()
